### ISSUE TITLE
ES6 in Sequelize codebase

### ISSUE THREAD

**Issue by overlookmotel** [sequelize/sequelize#3951]
@mickhansen recently raised the possibility of using [babel](https://babeljs.io/) to allow Sequelize to be written using ES6 features while retaining compatibility with versions of node earlier than v0.12.

@janmeier and @mbroadst were in favour, @BridgeAR has reservations.

The original discussion is here: https://github.com/sequelize/sequelize/pull/3929/files#r32521460

I thought it worthwhile to give this its own issue for discussion as it's a big topic.

**Comment by overlookmotel** [comment]
My opinion is:

I'm writing node apps in ES6. Using generators, `co` and `yield` makes it *sooooo* much simpler to write async code. No more endless `.then()` promise chains...

But I'm running these apps in node v0.12 with `--harmony` so the ES6 features are native.

I'd be broadly in favour of transitioning Sequelize to ES6, because the benefits in terms of writing shorter more comprehensible code are so large.

However, what are the implications for debugging? If the source code goes through a babel compilation, does that make it hard to locate the source of a bug, if a thrown error references a line number which doesn't correspond to the right line in the source code?

**Comment by overlookmotel** [comment]
I'd suggest if you wanted to go this way, to restrict ES6 features used to those available natively in node v0.12 with `--harmony`, so it can run without compiling on node v0.12. Babel compilation could be switched on only for older node versions.

**Comment by mickhansen** [comment]
@overlookmotel The beautiful dynamic object keys and structural deassignments aren't in harmony though :/

**Comment by mbroadst** [comment]
IMHO I think the suggestion to just use harmony is sort of dead in the water since you'd basically be telling users that sequelize no longer works with 0.10.x and if they want to use it with 0.12.x they have to run with ```--harmony```: no thanks.

I think babel is your only choice if you want to go down this path, it provides a nice platform for which to support all the things you need to support here while still looking forward.

**Comment by overlookmotel** [comment]
@mbroadst My suggestion was that Sequelize would work natively on 0.12.x with `--harmony`, but that when run on 0.10.x it would be babelized. So it'd run on both platforms, but with no need for babelization if you're on node 0.12.x.

But in any case, I didn't realize that as @mickhansen said some of the best ES6 features aren't available in 0.12.x harmony. So my whole idea is dead in the water for that reason!

My question remains though: What are the implications for debugging babelized code? Presumably the cross-compile changes line numbers, so when looking at an error stack trace, it could be hard to follow what's going on - the stack trace wouldn't correspond to the source code. This would be a deal-breaker for me as it could make Sequelize unmaintainable.

That said, I haven't tried babel, so there may be solution for this. Anybody know?

**Comment by mickhansen** [comment]
It would break stack traces, i'm not sure if node.js has great sourcemap support, likely not.

**Comment by overlookmotel** [comment]
I'd imagine it'd be possible for babel to be written so that line numbers are preserved (by making very long lines in some cases!), but I don't know if it does this.

**Comment by overlookmotel** [comment]
This blog post talks about using source maps with ES6 compilation:
http://www.2ality.com/2015/04/node-es6-transpiled.html

**Comment by mbroadst** [comment]
https://github.com/evanw/node-source-map-support ?

**Comment by SpainTrain** [comment]
#### $0.02

In my experience using sourcemaps in node (which does not include a project as large as Sequelify), the source maps support using the module linked by @mbroadst has been fine for any debugging I've needed to do.

The tech for ES6 dev is pretty good, so if I were maintaining a popular ES5 project, I would probably worry most about how it affects the core contributors and community.  A few examples:

* What do maintainers spend most of their time on?  Would the improvement in productivity be worth the investment in learning and creating the tooling to use ES6?
* How important are PRs from non-core contributors?  What is the culture of the community?  Would using ES6 in the project reduce the number of PRs and negatively affect the project?
* Fork vs. evolve - What are the trade offs of evolving the current project rather than creating a new ground-up ES6 porject (a la koa)?

I love writing in ES6, but for a mature OSS project, I suspect the considerations that have the most gravity are the non-technical ones.

Related: https://github.com/sequelize/sequelize/issues/3840

**Comment by cusspvz** [comment]
In my opinion all node LTS versions should be supported.

Regarding to `0.10` and `0.12` versions support I agree with @mbroadst , babel repl seems the only reasonable choice.

**Comment by xpepermint** [comment]
+1

**Comment by overlookmotel** [comment]
Now that Node v4 is released and so there are two versions of node (v.0.12 with harmony flag and v4) which support some ES6 features, I think it's worthwhile looking at this again.

@mickhansen has mentioned dynamic object keys and destructuring assignment, but for me the killer feature of ES6 is being able to use [co](https://www.npmjs.com/package/co) and generators. This makes it *so* much easier to write async code with is both faster to write and much more comprehensible. I've started using it everywhere and it's revolutionised my coding.

Example code is current Promises style:

```js
var create = function(options) {
  var self = this;
  return Promise.try(function() {
    if (options.hooks) return self.runHooks('beforeCreate', options);
  }).then(function() {
    return self.runQuery( /* ... etc ... */ );
  }).tap(function() {
    if (options.hooks) return self.runHooks('afterCreate', options);
  }).then(function(instance) {
    return instance;
  });
};
```

With co...

```js
var create = co.wrap(function*(options) {
  if (options.hooks) yield self.runHooks('beforeCreate', options);
  var instance = yield self.runQuery( /* ... etc ... */ );
  if (options.hooks) yield self.runHooks('afterCreate', options);
  return instance;
});
```

This is available natively in node v0.12 with `--harmony` flag without any transpiling. And also in node v4. To continue support for node v0.10 (which I think Sequelize should), it would require babel.

I see 3 options:

1. Keep everything as is, natively supporting node v0.10 onwards
2. Restrict Sequelize code to only what is natively provided in node v0.12 with `--harmony`, and transpile with babel for node v0.10
3. Go the whole hog, and transpile with babel for all versions of node, to allow all ES6 features (and also async/await from ES7)

Personally, I think option (2) is a good middle way allowing users of modern node to run Sequelize code natively, while still supporting node v0.10. Extra language features could also be brought into use (e.g. destructuring assignment) as node v4 matures and these ES6 features become natively supported.

But that's just my personal opinion...

**Comment by rickihastings** [comment]
Node 4 supports a lot of ES6 functionality now too, including native promises meaning a dependency on bluebird could be removed. This may come with a performance improvement too.

**Comment by BridgeAR** [comment]
@rickihastings bluebird is [faster](http://bluebirdjs.com/docs/benchmarks.html), uses less memory and has a lot more features than native promises. Also some that sequelize use, like `finally` and long stack traces.

**Comment by mbroadst** [comment]
@rickihastings incorrect. bluebird is likely to always be more performant than the native implementations due to the fact that it can "cheat". See [this](http://bluebirdjs.com/docs/benchmarks.html), and [this](http://programmers.stackexchange.com/questions/278778/why-are-native-es6-promises-slower-and-more-memory-intensive-than-bluebird)

**Comment by rickihastings** [comment]
Ahh fair enough, did not know that! I've been using native promises in favour, guess I'll go back to bluebird.

**Comment by overlookmotel** [comment]
@mbroadst Interesting stuff!

**Comment by mickhansen** [comment]
We're leaning towards using ES6/Babel with bluebird coroutines in Sequelize v4.0.0.
(maybe not bluebird coroutines, need to figure out the performance impact).

**Comment by overlookmotel** [comment]
Just to let you know, bluebird co-routines don't support CLS (or at least they didn't in bluebird v2).

The maintainers of `co` didn't seem interested in adding the ability to use other promise implementations (see https://github.com/tj/co/pull/226), so I released my own modules [co-use](https://www.npmjs.com/package/co-use) and [co-bluebird](https://www.npmjs.com/package/co-bluebird) that repllicate co's functionality but using bluebird or your choice of another promise implementation.

In my own code, I use the following to get CLS support:

```js
var co = require('co-use).use(Sequelize.Promise);
```

But... I have run into problems where CLS context gets lost on occasion. I haven't had time to get to the bottom of why this is happening.

My suspicion is that there are some problems in Sequelize's shims of bluebird to maintain CLS context. In my opinion best approach is to sort out the bluebird/CLS problem and publish that as a separate module which then Sequelize can depend on.

I've got about a month left on an all-consuming project, but after that hope to have time to investigate this further and find a solution. I'm using CLS a lot in for my own apps, and need it to be rock solid, so I may be able to help with this effort. But just to let you know that the path to migrating to co-routines isn't as smooth as you might hope.

**Comment by overlookmotel** [comment]
@mickhansen What's expected timeline for Sequelize v4 being released?

**Comment by mickhansen** [comment]
While i'd love to switch to bluebird coroutines i don't believe we'll make that switch just yet. While we'll officially drop CI support for 0.10 i won't actively stop people from trying by using generators.
Also still unsure of the perf hit using generators.

**Comment by mickhansen** [comment]
@overlookmotel Not sure, need to identify everything i want to change first :)

**Comment by overlookmotel** [comment]
Sequelize v4: so we're talking a couple of months maybe?

Performance hit: a friend of mine did a rough benchmark of co-routines (using my [co-bluebird](https://www.npmjs.com/package/co-bluebird)) vs bluebird promises and found a negligible performance cost. In some cases co-routines appeared to improve performance! But that makes no sense, so I think it should be taken with a pinch of salt. It needs more careful analysis.

**Comment by mickhansen** [comment]
Accounting for history in our otherwise optimistic estimations, yes probably a couple of months.
I think we'll skip bluebird-coroutines for now, having ES6 stuff for productivity will be a step in the right direction in anycase.
But the first step is identifying everything we want to break for producitivity and APIs sake.

**Comment by overlookmotel** [comment]
As a stepping stone towards co-routines, may I suggest using [promisify-any](https://www.npmjs.com/package/promisify-any) for hooks? That'd allow Sequelize users to write their hooks as co-routines if they wish.

**Comment by mickhansen** [comment]
@overlookmotel With a few changes that might work (looks like its using it own version of bluebird for once), but it should be as simple as checking whether or not it's a generator function and just wrapping it in http://bluebirdjs.com/docs/api/promise.coroutine.html if it is.

**Comment by mickhansen** [comment]
```js
Model.beforeCreate(async function () {

});
```

With babel and bluebird-coroutines on the user side would work out of the box in any case.

**Comment by overlookmotel** [comment]
Having switched to co-routines a while back, I really can't sing their praises enough. It makes code *so* much easier, faster and more pleasurable to write.

I've found I produce less buggy code as you don't have to worry about all the code inside `.then(function() { /* code here */ })` constructs having a different scope, and value for `this` than the next `.then()` block.

It's a beautiful thing man!

**Comment by mickhansen** [comment]
I use them extensively in all my code aswell, i'm just not sure it would be responsible to switch our entire underlying library to use it yet :)

**Comment by janmeier** [comment]
I'd hate having to switch, but bluebird has built in support for domains - which also allow you to store values on them, and maintain callback context, see https://github.com/petkaantonov/bluebird/issues/583

**Comment by mickhansen** [comment]
domains have been deprecated forever though: https://nodejs.org/api/domain.html

**Comment by janmeier** [comment]
So basically there is no universally supported method for callback context ...

**Comment by overlookmotel** [comment]
@janmeier Well CLS *should* work, but bluebird author has rejected suggestions in the past to include CLS support. The shim of bluebird used in Sequelize to add CLS support is I suspect a little buggy, so that's what's meaning we don't have full bullet-proof CLS support.

But, *in theory*, CLS *is* a universal method for callback context. Just got to iron out the bugs of integrating it with bluebird.

**Comment by Verdier** [comment]
@overlookmotel @mickhansen there is absolutely no need  to use `promisify-any` since `Promise` is the way async functions will work tomorrow. If you want to use  generators with today Sequelize, you juste have to do something like that:

```js
beforeValidate: co.wrap(function *() { ... })
```
and tomorrow (has said by @mickhansen):

```js
beforeValidate: async function() { ... }
```

both of the solutions return a `Promise`. Even `koa` has drop generators support in favore of `co.wrap` or `async` used by the library consumer.

**Comment by MarkMurphy** [comment]
I'm surprised no one has mentioned async/await here. 

```js
myAsyncTask().then(function (result) {
  return result.something()
}).then(function (doc) {
  console.log(doc)
}).catch(function (err) {
  console.log(err);
});
```

becomes:

```js
try {
  let result = await myAsyncTask()
  let doc = await result.something()
  console.log(doc);
} catch (err) {
  console.log(err);
}
```

have a look [here](http://pouchdb.com/2015/03/05/taming-the-async-beast-with-es7.html).

**Comment by mickhansen** [comment]
@MarkMurphy It has actually been mentioned :) We're not going with it internally since it would require bluebird-coroutines are we're not sure of the perf implications.

However users are as always welcome to us it, and we might update docs to reflect it.

**Comment by overlookmotel** [comment]
@Verdier I just suggested using [promisify-any](https://www.npmjs.com/package/promisify-any) as currently hooks can be written either as classic node callback functions or promise-returning functions. [promisify-any](https://www.npmjs.com/package/promisify-any) takes care of that, so the code handling callback functions could be removed from the Sequelize codebase.

The additional support for generator co-routines is a bonus.

**Comment by overlookmotel** [comment]
@Verdier There are also some died-in-the-wool callback devotees who like that style of programming and will probably never change. Seems bonkers to me, but different strokes for different folks... our vision of the future isn't necessarily shared by everyone.

**Comment by Verdier** [comment]
Yep, @overlookmotel I agree!

**Comment by mickhansen** [comment]
I say we remove callback support though: https://github.com/sequelize/sequelize/issues/5228

**Comment by sushantdhiman** [comment]
Following is the summary for ES6 adoption in Sequelize codebase

- Supporting only Node v4 and up for upcoming `Sequelize@4.0`
- No `Babel` or `TypeScript` will be used. We want Sequelize to run directly after git clone without any transpilation overhead
- Use of only default enabled ES6 feature from Node v4 , so we have `template strings`, `Class` and `arrow functions`
- We may adopt `co routines` #6116 , if CLS support is guaranteed 

Discussion moved to #5898
